compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java

```diff
@@ -474,4 +487,58 @@ private void checkUniqueness(BLangConstant constant) {
             }
         }
     }
+
+    private void updateSymbolType(BLangConstant constant) {
+        if (constant.symbol.kind == SymbolKind.CONSTANT && constant.symbol.type.getKind() != TypeKind.FINITE &&
+                constant.symbol.value != null) {
+            BFiniteType finiteType = checkType(constant, constant.symbol.value.value, constant.symbol.type,
+                    constant.symbol.pos);
+            if (finiteType != null) {
+                constant.symbol.type = finiteType;
+            }
+        }
+    }
+
+    public BFiniteType createFiniteType(BLangConstant constant, BLangExpression expr) {
+        BTypeSymbol finiteTypeSymbol = Symbols.createTypeSymbol(SymTag.FINITE_TYPE,
+                Flags.asMask(EnumSet.noneOf(Flag.class)), Names.EMPTY, constant.symbol.pkgID, null,
+                constant.symbol.owner, constant.symbol.pos, VIRTUAL);
+        BFiniteType finiteType = new BFiniteType(finiteTypeSymbol);
+        finiteType.addValue(expr);
+        return finiteType;
+    }
+
+    private BFiniteType checkType(BLangConstant constant, Object value, BType type, Location pos) {
+        switch (type.getKind()) {
+            case INT:
+            case BYTE:
+            case FLOAT:
+            case DECIMAL:
+                return createFiniteType(constant, createConstantNumericLiteralExpression(value, type, pos));
+            case STRING:
+            case NIL:
+            case BOOLEAN:
+                return createFiniteType(constant, createConstantLiteralExpression(value, type, pos));
+            default:
+                return null;
+        }
+    }
+
+    private BLangNumericLiteral createConstantNumericLiteralExpression(Object value, BType type, Location pos) {
+        BLangNumericLiteral literal = (BLangNumericLiteral) TreeBuilder.createNumericLiteralExpression();
+        literal.value = value;
+        literal.isConstant = true;
+        literal.setBType(type);
+        literal.pos = pos;
+        return literal;
+    }
+
+    private BLangLiteral createConstantLiteralExpression(Object value, BType type, Location pos) {
+        BLangLiteral literal = (BLangLiteral) TreeBuilder.createLiteralExpression();
+        literal.value = value;
+        literal.isConstant = true;
+        literal.setBType(type);
+        literal.pos = pos;
+        return literal;
```

Comment:
Common logic can be extracted out to a separate method?

Reply:
Yeah. I have missed it. Thanks.